src/nvim/optionstr.c

```diff
@@ -1905,35 +1906,42 @@ char *did_set_string_option(int opt_idx, char **varp, char *oldval, char *errbuf
     }
 
     // Trigger the autocommand only after setting the flags.
-    if (varp == &(curbuf->b_p_syn)) {
-      do_syntax_autocmd(curbuf, value_changed);
-    } else if (varp == &(curbuf->b_p_ft)) {
-      do_filetype_autocmd(curbuf, varp, opt_flags, value_changed);
-    } else if (varp == &(curwin->w_s->b_p_spl)) {
-      did_set_spelllang_source(curwin);
+    if (varp == &(buf->b_p_syn)) {
+      do_syntax_autocmd(buf, value_changed);
+    } else if (varp == &(buf->b_p_ft)) {
+      do_filetype_autocmd(buf, varp, opt_flags, value_changed);
+    } else if (varp == &(win->w_s->b_p_spl)) {
+      did_set_spelllang_source(win);
     }
   }
 
   if (varp == &p_mouse) {
-    setmouse();  // in case 'mouse' changed
+    setmouse_for(buf, win);  // in case 'mouse' changed
```

Comment:
This is wrong. 'mouse' is a global option and `setmouse()` should always only use current window/buffer.